File: hestiaGO/hestiaTESTING/Conclude_test.go
// Copyright 2022 "Holloway" Chew, Kean Ho <kean.ho.chew@zoralab.com>
// Copyright 2022 ZORALab Enterprise <tech@zoralab.com>
//
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy of
// the License at
//
//                  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations under
// the License.

package hestiaTESTING

import (
	"testing"
)

func _testConcludeScenarios() []*Scenario {
	return []*Scenario{
		{
			Description: `
Test Conclude() is able to work properly with proper Scenario settings.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_PROPER_SWITCHES,
				cond_PROPER_VERDICT,
			},
		}, {
			Description: `
Test Conclude() is able to work properly with empty Name setting.
`,
			Switches: []string{
				cond_EMPTY_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_PROPER_SWITCHES,
				cond_PROPER_VERDICT,
			},
		}, {
			Description: `
Test Conclude() is able to work properly with empty Switches setting.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_EMPTY_SWITCHES,
				cond_PROPER_VERDICT,
			},
		}, {
			Description: `
Test Conclude() is able to work properly with nil Switches setting.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_NIL_SWITCHES,
				cond_PROPER_VERDICT,
			},
		}, {
			Description: `
Test Conclude() is able to work properly with empty log setting.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_EMPTY_LOG,
				cond_PROPER_SWITCHES,
				cond_PROPER_VERDICT,
			},
		}, {
			Description: `
Test Conclude() is able to work properly with nil log setting.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_NIL_LOG,
				cond_PROPER_SWITCHES,
				cond_PROPER_VERDICT,
			},
		}, {
			Description: `
Test Conclude() is able to work properly with empty description setting.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_EMPTY_DESCRIPTION,
				cond_NIL_LOG,
				cond_PROPER_SWITCHES,
				cond_PROPER_VERDICT,
			},
		}, {
			Description: `
Test Conclude() is able to panic when nil Scenario is supplied.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_PROPER_SWITCHES,
				cond_SUPPLY_NIL_SCENARIO,
				cond_PROPER_VERDICT,
				expect_PANIC,
			},
		}, {
			Description: `
Test Conclude() is able to panic when verdict is an unknown.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_PROPER_SWITCHES,
				cond_UNKNOWN_VERDICT,
				expect_PANIC,
			},
		}, {
			Description: `
Test Conclude() is able to work properly when verdict is VERDICT_FAIL.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_PROPER_SWITCHES,
				cond_FAIL_VERDICT,
			},
		}, {
			Description: `
Test Conclude() is able to work properly when verdict is VERDICT_SKIP.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_PROPER_SWITCHES,
				cond_SKIP_VERDICT,
			},
		},
	}
}

func TestConcludeAPI(t *testing.T) {
	scenarios := _testConcludeScenarios()

	for i, s := range scenarios {
		s.ID = uint64(i)
		s.Name = suite_CONCLUDE_API

		// prepare
		ts := &Scenario{}
		testlib_ConfigureName(s, ts)
		testlib_ConfigureDescription(s, ts)
		testlib_ConfigureLog(s, ts)
		testlib_ConfigureSwitches(s, ts)
		ts.verdict = value_DEFAULT_VERDICT
		verdict := testlib_GenerateVerdict(s)

		// test
		_panick := Exec(func() any {
			if !HasCondition(s, cond_SUPPLY_NIL_SCENARIO) {
				Conclude(ts, verdict)
			} else {
				Conclude(nil, verdict)
			}
			return ""
		})
		panick := _panick.(string)

		// log output
		Log(s, _renderString("Test Scenario's Verdict	= %#v", ts.verdict))
		Log(s, _renderString("Got Panic			= %q", panick))

		// assert
		if !testlib_AssertPanic(s, panick) {
			Conclude(s, VERDICT_FAIL)
			t.Fail()
		}

		if !testlib_AssertConclude(s, ts) {
			Conclude(s, VERDICT_FAIL)
			t.Fail()
		}

		if Conclusion(s) != VERDICT_FAIL {
			Conclude(s, VERDICT_PASS)
		}

		// report
		t.Logf("\n%s\n\n\n", ToString(s))
	}
}

File: hestiaRUST/hestia_testing/conclude_test.rs
// Copyright 2022 "Holloway" Chew, Kean Ho <kean.ho.chew@zoralab.com>
// Copyright 2022 ZORALab Enterprise <tech@zoralab.com>
//
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy of
// the License at
//
//                  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations under
// the License.
use crate::hestia_testing;
use crate::hestia_testing::testlibs_test;

// test suites
const SUITE_NAME: &str = "hestia_testing::conclude API";

// test conditions

// test values

// test libs
fn assert_panic(s: &hestia_testing::Scenario) -> bool {
	if hestia_testing::has_condition(s, testlibs_test::COND_PROPER_VERDICT) {
		return true;
	}

	if hestia_testing::has_condition(s, testlibs_test::COND_FAIL_VERDICT) {
		return true;
	}

	if hestia_testing::has_condition(s, testlibs_test::COND_SKIP_VERDICT) {
		return true;
	}

	if hestia_testing::has_condition(s, testlibs_test::COND_UNKNOWN_VERDICT) {
		return false;
	}

	if hestia_testing::has_condition(s, testlibs_test::COND_INITIAL_VERDICT) {
		return false;
	}

	return false;
}

fn assert_output(s: &hestia_testing::Scenario, output: hestia_testing::Error) -> bool {
	if hestia_testing::has_condition(s, testlibs_test::COND_PROPER_VERDICT) {
		return output == hestia_testing::ERROR_OK;
	}

	if hestia_testing::has_condition(s, testlibs_test::COND_FAIL_VERDICT) {
		return output == hestia_testing::ERROR_OK;
	}

	if hestia_testing::has_condition(s, testlibs_test::COND_SKIP_VERDICT) {
		return output == hestia_testing::ERROR_OK;
	}

	if hestia_testing::has_condition(s, testlibs_test::COND_UNKNOWN_VERDICT) {
		return output != hestia_testing::ERROR_OK;
	}

	if hestia_testing::has_condition(s, testlibs_test::COND_INITIAL_VERDICT) {
		return output != hestia_testing::ERROR_OK;
	}

	return false;
}

fn test_conclude_algorithm(id: u64, desc: String, switches: Vec<String>) {
	// prepare
	let mut s: &mut hestia_testing::Scenario = &mut hestia_testing::new_scenario();
	s.id = id;
	s.name = String::from(SUITE_NAME);
	s.description = desc;
	s.switches = switches;
	hestia_testing::conclude(s, hestia_testing::VERDICT_PASS);

	// test
	let ts: &mut hestia_testing::Scenario = &mut hestia_testing::new_scenario();
	let verdict: hestia_testing::Verdict = testlibs_test::create_verdict(s);
	hestia_testing::log(s, format!("Given verdict	: '''\n{}\n'''", verdict));

	if hestia_testing::has_condition(s, testlibs_test::COND_UNKNOWN_VERDICT)
		|| hestia_testing::has_condition(s, testlibs_test::COND_INITIAL_VERDICT)
	{
		println!("{}", hestia_testing::to_string(s));
	}
	let output: hestia_testing::Error = hestia_testing::conclude(ts, verdict);

	hestia_testing::log(s, format!("Got output	: '''\n{}\n'''", output));

	// assert
	if !assert_panic(s) {
		hestia_testing::conclude(s, hestia_testing::VERDICT_FAIL);
	}

	if !assert_output(s, output) {
		hestia_testing::conclude(s, hestia_testing::VERDICT_FAIL);
	}

	// report
	if !(hestia_testing::has_condition(s, testlibs_test::COND_UNKNOWN_VERDICT)
		|| hestia_testing::has_condition(s, testlibs_test::COND_INITIAL_VERDICT))
	{
		println!("{}", hestia_testing::to_string(s));
	}
	assert!(hestia_testing::conclusion(s) == hestia_testing::VERDICT_PASS);
}

// test suites
hestia_testing_exec!(test_conclude_initial_verdict, true, {
	test_conclude_algorithm(
		4,
		"\
test hestia_testing::conclude() is able to panic with initial (0) verdict value.
"
		.to_string(),
		vec![String::from(testlibs_test::COND_INITIAL_VERDICT)],
	)
});

hestia_testing_exec!(test_conclude_unknown_verdict, true, {
	test_conclude_algorithm(
		3,
		"\
test hestia_testing::conclude() is able to panic with unknown verdict value.
"
		.to_string(),
		vec![String::from(testlibs_test::COND_UNKNOWN_VERDICT)],
	)
});

hestia_testing_exec!(test_conclude_skip_verdict, {
	test_conclude_algorithm(
		2,
		"\
test hestia_testing::conclude() is able to run with VERDICT_SKIP.
"
		.to_string(),
		vec![String::from(testlibs_test::COND_SKIP_VERDICT)],
	)
});

hestia_testing_exec!(test_conclude_fail_verdict, {
	test_conclude_algorithm(
		1,
		"\
test hestia_testing::conclude() is able to run with VERDICT_FAIL.
"
		.to_string(),
		vec![String::from(testlibs_test::COND_FAIL_VERDICT)],
	)
});

hestia_testing_exec!(test_conclude_proper_verdict, {
	test_conclude_algorithm(
		0,
		"\
test hestia_testing::conclude() is able to run with VERDICT_PASS.
"
		.to_string(),
		vec![String::from(testlibs_test::COND_PROPER_VERDICT)],
	)
});
